### PR TITLE
streamingest: write and manage PTS on the destination tenant

### LINKED ISSUE #92093: c2c: write a protected timestamp on the destination cluster
Data is streamed from the source to the destination tenant over several partitions. A frontier timestamp that is updated during ingestion, tracks the highest timestamp up to which all partitions on the destination cluster have completed streaming. As time progresses and more data is ingested, versions that fall below the destination tenant's GCThreshold will become eligible for GC. Timestamps below the tenant's GCThreshold become ineligible for cutover since those version are no longer available to revert to.

To ensure that the user always has some generous window of timestamps to cutover to, we must write a protected timestamp over the ingesting tenant's keyspan that protects at `FrontierTimestamp - C2CTTL`, where C2CTTL can default to 25hours to begin with. This can later be made a configurable value that is specified when the replication stream is being setup. As the frontier moves forward, this protected timestamp must be pulled up as well so as to make data outside the C2CTTL window eligible for GC.

Jira issue: CRDB-21570

Epic: CRDB-18749

### PR DESCRIPTION
**Pull request by adityamaru**
During C2C replication as the destination tenant is ingesting KVs we must protect a certain window of MVCC revisions from garbage collection so that the user can cutover to any of the timestamps that lie within this window.

To this effect we introduce a `ReplicationTTLSeconds` field to the replication job payload that governs the size of this window relative to the replication job's highwatermark (frontier timestamp). On the first resumption of the replication job we write a protected timestamp record on the destination tenant's keyspace protecting all revisions above `now()`. As the replication job updates its highwatermark, the PTS record is pulled up to protect above `highWatermark - ReplicationTTLSeconds`. This active management of the PTS always ensures that users can cutover to any time in (highWatermark-ReplicationTTLSeconds, highWatermark] and older revisions are gradually made eligible for GC as the frontier progresses.

The PTS is released if the replication job fails or is cancelled.

Fixes: #92093

Release note: None

### CODE REVIEW

File: pkg/sql/gcjob/refresh_statuses.go
Comment (stevendanna) [inline pkg/sql/gcjob/refresh_statuses.go:468]: I think we can nuke `tenantHasActiveReplicationJob` as well.

We _may_ want to beef up the tests around this garbage collection waiting, but that doesn't need to happen in this PR.

File: pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go
Comment (adityamaru) [reply pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go:429]: indeeed, nice catch. I added to the test to also ensure a complete replication job clears up the PTS. Is the test doing too much? Its definitely a slow test taking 45s to run all 4 combinations.

File: pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go
Comment (adityamaru) [reply pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go:578]: It does, I'll wait for #92320 to merge and rebase this before merging.

File: pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go
Comment (adityamaru) [reply pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go:377]: Oh I didn't realize this. I wonder if we need to refactor `Resume` to run a different code path if it's being resumed after a pause due to failure. This way we minimize what we do between resumption and checking for cutover.

**Review by stevendanna — APPROVED** [review]

File: pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go
Comment (adityamaru) [reply pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go]: done!

**Comment by adityamaru** [comment]
TFTR!

bors r=stevendanna

**Review by lidorcarmel — APPROVED** [review]
> The PTS is released if the replication job fails or is cancelled.

1. do we allow cancelling the job? I think we said we don't, instead we only let you drop the dest tenant, right?
2. failure is interesting, do we think we can resume the job? if not, then maybe we should keep the PTS? after all, there is no data coming into this cluster anyway, so why not freeze everything until the tenant is dropped or the job is restarted?

this and my other comments are not critical so please keep merging..

File: pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go
Comment (adityamaru) [reply pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go:311]: This is a valid concern but I'm not sure it's one the replication job should be responsible for solving. Once the cutover is successful the replication job ceases to exist and we will go back to respecting the span config that applies to the tenant keyspace. In an ideal work we make the uncorking of GC not as chaos-inducing as it is today, but if the user does not wish to have a flurry of GC activity they can either:

1. Configure the GC TTL on the tenant keyspan to be 25h so that even after cutover we do not suddenly have a large amount of data eligible for GC.

2. Run the replication stream with a smaller retention window once we make this `ReplicationTTLSeconds` a user-configurable value.

I do agree that this is something we will need to document explicitly.

Comment (lidorcarmel) [reply pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go:311]: good points! thanks.

**Comment by adityamaru** [comment]
> do we allow cancelling the job?

I believe we do at least for now. `CANCEL JOB <replication-job-id>` would cancel the replication job.

> I think we said we don't, instead we only let you drop the dest tenant, right?

That's the desired UX, but `DROP TENANT` internally cancels the replication job associated with the tenant being dropped, so we still need to release the PTS as part of the `OnFailOrCancel` method invoked on job cancellation. If we do not release the PTS then the GC job that is queued as part of the `DROP TENANT` to clean up the tenant's keyspace will forever wait for the protected timestamp to be removed.

> failure is interesting, do we think we can resume the job? if not, then maybe we should keep the PTS? after all, there is no data coming into this cluster anyway, so why not freeze everything until the tenant is dropped or the job is restarted?

Yup, you're right we don't actually fail the job - https://github.com/cockroachdb/cockroach/blob/0015d5d0dbbda9b810915c89778e31103aee7a39/pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go#L372. All errors during replication job execution are "handled" and the replication job is moved into a paused state. This means that the `OnFailOrCancel` hook will not be run as the job goes from running to paused on failure. Consequently the PTS is not released and so the user can resume the replication job and will find the PTS still present and holding up GC.

File: pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go
Comment (adityamaru) [reply pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go:405]: If we return an error here then that will bubble up to `OnFailOrCancel` which is a method the job registry does not allow to fail. In other words, we will retry `OnFailOrCancel` endlessly (with backoff) until it succeeds which it will never since it won't be able to find the record. A reason why we might not find the record is if `OnFailOrCancel` gets retried causing us to invoke this method more than once in separate txns. The expectation is that all operations in `Resume` and `OnFailOrCancel` should be idempotent and so we safeguard against multiple invocations.

Comment (lidorcarmel) [reply pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go:405]: oh this definitely makes sense in that case. thanks for explaining!

**Pull request merged**
